README.md
# Paystack Flutter

This plugin provides an easy way to receive payments on Android and iOS apps with Paystack. It uses the native [Android](https://github.com/PaystackHQ/paystack-android) and [iOS libraries](https://github.com/PaystackHQ/paystack-ios) under the hood and provides a unified API for initializing payment in a platform-agnostic way. The flow surroudning how Paystack payments work is well written in the [Android library documentation](https://github.com/PaystackHQ/paystack-android), so we'll just skip all the formalities and demonstrate how to use this library.

## Usage

### Step 1 - Add this plugin as a dependency to your flutter project

The good folks at [Flutter explains how here](https://flutter.io/platform-plugins/)

### Step 2 - Accept payment

This step assumes you've already built your UI for accepting card details from your application user. And of course, you have your Paystack public API key.

Import the plugin in the file where you want to accept payments.

```
// the plugin exposes a Card class that conflicts with another class of the same name in Flutter's material plugin.
// Importing as PaystackSDK allows you to access the Card class via PaystackSDK.Card
import 'package:paystack_flutter/paystack_flutter.dart' as PaystackSDK;
```

Next, initialize Paystack by proividing your public API key. You should preferably do this once, when the page loads and the public key value will remain set. You can subsequently use the SDK to receive payments multiple times in the page.

```
Future<void> initPaystack() async {
    String paystackKey = "pk_test_xxxxxxxxxxxxxxx";
    try {
        await PaystackSDK.Paystack.initialize(paystackKey);
        // Paystack is ready for use in receiving payments
    } on PlatformException {
        // well, error, deal with it
    }
}
```

Receive payments already!

```
initPayment() {
    // pass card number, cvc, expiry month and year to the Card constructor function
    var card = PaystackSDK.Card("5060666666666666666", "123", 12, 2020);

    // create a transaction with the payer's email and amount (in kobo)
    var transaction = PaystackSDK.Transaction("wisdom.arerosuoghene@gmail.com", 100000);
    
    // debit the card (using Javascript style promises)
    transaction.chargeCard(card)
      .then((transactionReference) {
        // payment successful! You should send your transaction request to your server for validation
      })
      .catchError((e) {
        // oops, payment failed, a readable error message should be in e.message 
      });
}
```
## Contributing

Contributions are most welcome. You could improve this documentation, add method to support more Paystack features or just clean up the code. Just do your thing and create a PR. This started out as a quick work to achieve paystack payments on Android and iOS. A lot could have been done better.

CHANGELOG.md
## 0.0.1

## Features

- Accept payment with email, amount and card details


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:paystack_flutter_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


ios/Classes/SwiftPaystackFlutterPlugin.swift
import Flutter
import UIKit
import Paystack

public class SwiftPaystackFlutterPlugin: NSObject, FlutterPlugin {
    private static var CHARGE_CARD_ERROR_CODE = "chargeCard"
    private static var INITIALIZE_PAYSTACK_ERROR_CODE = "initializePaystack"
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "paystack_flutter", binaryMessenger: registrar.messenger())
    let instance = SwiftPaystackFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initializePaystack":
        self.initializePaystack(call, result)
        
    case "chargeCard":
        self.chargeCard(call, result)
        
    default:
        result(FlutterError(code: "UNKNOWN METHOND",
                            message: "Request not understood",
                            details: nil))
    }   
  }
    
    private func initializePaystack(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        if call.arguments == nil {
            result(FlutterError(code: SwiftPaystackFlutterPlugin.INITIALIZE_PAYSTACK_ERROR_CODE,
                                message: "initializePaystack requires paystackKey argument",
                                details: nil))
            return
        }
        
        let publicKey = call.arguments as! String
        Paystack.setDefaultPublicKey(publicKey)
        result(nil)
    }
    
    private func chargeCard(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        if call.arguments == nil || !(call.arguments is NSDictionary) {
            result(FlutterError(code: SwiftPaystackFlutterPlugin.CHARGE_CARD_ERROR_CODE,
                                message: "chargeCard requires card and transaction information",
                                details: nil))
            return
        }
        
        let cardAndTransactionDictionary = call.arguments as! NSDictionary
        let email = cardAndTransactionDictionary.value(forKey: "email") as? String
        let amount = cardAndTransactionDictionary.value(forKey: "amount") as? Int
        let cardNumber = cardAndTransactionDictionary.value(forKey: "cardNumber") as? String
        let cvc = cardAndTransactionDictionary.value(forKey: "cvc") as? String
        let expiryMonth = cardAndTransactionDictionary.value(forKey: "expiryMonth") as? Int
        let expiryYear = cardAndTransactionDictionary.value(forKey: "expiryYear") as? Int
        
        if email == nil || amount == nil || cardNumber == nil || cvc == nil || expiryMonth == nil || expiryYear == nil {
            result(FlutterError(code: SwiftPaystackFlutterPlugin.CHARGE_CARD_ERROR_CODE,
                                message: "1 or more required card/transaction information not provided",
                                details: nil))
            return
        }
        
        let viewController = UIApplication.shared.keyWindow?.rootViewController
        if viewController == nil {
            result(FlutterError(code: SwiftPaystackFlutterPlugin.CHARGE_CARD_ERROR_CODE,
                                message: "Paystack payments can only be initialized from a View Controller",
                                details: nil))
            return
        }
        
        let cardParams = PSTCKCardParams.init();
        cardParams.number = cardNumber!
        cardParams.cvc = cvc!
        cardParams.expYear = UInt(expiryYear!)
        cardParams.expMonth = UInt(expiryMonth!)
        
        let transactionParams = PSTCKTransactionParams.init();
        transactionParams.amount = UInt(amount!)
        transactionParams.email = email!
        
        func errorHandler(error: Error, reference: String?) -> Void {
            result(FlutterError(code: SwiftPaystackFlutterPlugin.CHARGE_CARD_ERROR_CODE,
                                message: error.localizedDescription,
                                details: reference))
        }
        func validationRequestedHandler(reference: String) -> Void {
            // todo, use a stream (?) to keep user updated for this kind of scenario
            // but remember the process still continues until onSuccess or onError is called
        }
        func successHandler(reference: String) -> Void {
            result(reference)
        }
        
        PSTCKAPIClient.shared().chargeCard(cardParams, forTransaction: transactionParams, on: viewController!,
                                           didEndWithError: errorHandler,
                                           didRequestValidation: validationRequestedHandler,
                                           didTransactionSuccess: successHandler)
    }
}


android/src/main/java/tech/itswisdomagain/paystackflutter/PaystackFlutterPlugin.java
package tech.itswisdomagain.paystackflutter;

import android.app.Activity;

import java.util.Map;

import co.paystack.android.Paystack;
import co.paystack.android.PaystackSdk;
import co.paystack.android.Transaction;
import co.paystack.android.model.Card;
import co.paystack.android.model.Charge;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** PaystackFlutterPlugin */
public class PaystackFlutterPlugin implements MethodCallHandler {
  private Activity activity;
  private static final String CHARGE_CARD_ERROR_CODE = "chargeCard";
  private static final String INITIALIZE_PAYSTACK_ERROR_CODE = "initializePaystack";

  private PaystackFlutterPlugin(Activity activity) {
    this.activity = activity;
  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    PaystackFlutterPlugin paystackFlutterPlugin = new PaystackFlutterPlugin(registrar.activity());

    MethodChannel channel = new MethodChannel(registrar.messenger(), "paystack_flutter");
    channel.setMethodCallHandler(paystackFlutterPlugin);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method) {
      case "initializePaystack":
        this.initializePaystack(call, result);
        break;

      case "chargeCard":
        this.chargeCard(call, result);
        break;

      default:
        result.notImplemented();
    }
  }

  private void initializePaystack(MethodCall call, Result result) {
    if (call.arguments == null) {
      result.error(INITIALIZE_PAYSTACK_ERROR_CODE, "initializePaystack requires paystackKey argument", null);
      return;
    }

    String publicKey = call.arguments();
    PaystackSdk.initialize(this.activity);
    PaystackSdk.setPublicKey(publicKey);
    result.success(null);
  }

  private void chargeCard(MethodCall call, final Result result) {
    // fetch the required card and transaction information from call.argument map
    if (call.arguments == null || !(call.arguments instanceof Map)) {
      result.error(CHARGE_CARD_ERROR_CODE, "chargeCard requires card and transaction information", null);
      return;
    }

    String email = call.argument("email");
    Integer amount = call.argument("amount");
    String cardNumber = call.argument("cardNumber");
    String cvc = call.argument("cvc");
    Integer expiryMonth = call.argument("expiryMonth");
    Integer expiryYear = call.argument("expiryYear");

    if (email == null || amount == null || cardNumber == null || cvc == null || expiryMonth == null || expiryYear == null) {
      result.error(CHARGE_CARD_ERROR_CODE, "1 or more required card/transaction information not provided", null);
      return;
    }

    Card card = new Card(cardNumber, expiryMonth, expiryYear, cvc);
    if (!card.isValid()) {
      result.error(CHARGE_CARD_ERROR_CODE, "card information is incorrect or invalid", null);
      return;
    }

    Charge charge = new Charge();
    charge.setCard(card);
    charge.setAmount(amount);
    charge.setEmail(email);

    PaystackSdk.chargeCard(this.activity, charge, new Paystack.TransactionCallback() {
      @Override
      public void onSuccess(Transaction transaction) {
        result.success(transaction.getReference());
      }

      @Override
      public void beforeValidate(Transaction transaction) {
        // todo, use a stream (?) to keep user updated for this kind of scenario
        // but remember the process still continues until onSuccess or onError is called
      }

      @Override
      public void onError(Throwable error, Transaction transaction) {
        result.error(CHARGE_CARD_ERROR_CODE, error.getLocalizedMessage(), transaction.getReference());
      }
    });
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:paystack_flutter/paystack_flutter.dart' as PaystackSDK;

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';
  bool _paymentReady = false;

  @override
  void initState() {
    super.initState();
    initPaystack();
  }

  // Platform messages may fail, so we use a try/catch PlatformException.
  Future<void> initPaystack() async {
    displayMessage("Initializing Paystack");
    
    String paystackKey = "pk_test_xxxxxxxxxxxxxxx";
    try {
      await PaystackSDK.Paystack.initialize(paystackKey);
      _paymentReady = true;
      displayMessage("Paystack Initialized. Ready to receive payment");
    } on PlatformException {
      displayError("Error intializing Paystack key to $paystackKey");
    }
  }

  initPayment() {
    if (!_paymentReady) {
      return;
    }

    displayMessage("Starting transaction...");
    _paymentReady = false;

    /*
    Test OTP Card from Paystack
    50606 66666 66666 6666
    CVV: 123
    PIN: 1234
    TOKEN: 123456
    EXPIRY: any date in the future
    */
    var card = PaystackSDK.Card("5060666666666666666", "123", 12, 2020);
    var transaction = PaystackSDK.Transaction("wisdom.arerosuoghene@gmail.com", 100000); // amount in kobo
    
    transaction.chargeCard(card)
      .then((transactionReference) {
        displayMessage("Transaction complete\nReference: $transactionReference");
        _paymentReady = true;
      })
      .catchError((e) {
        displayError(e.message);
        _paymentReady = true;
      });
  }

  displayMessage(String message) {
    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = message;
    });
  }

  displayError(String error) {
    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = 'Error:\n$error';
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Paystack Flutter SDK Sample'),
        ),
        body: Center(
          child: Text(_platformVersion),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: initPayment,
          tooltip: 'Init Payment',
          child: Icon(Icons.payment),
        ),
      ),
    );
  }
}
